File: Algoritmos Estudados/QuickSort.java
import java.io.*;
import java.util.*;

public class QuickSort {
    public static void main(String args[]) throws IOException {
        File file = new File("input.txt");
        BufferedReader br = new BufferedReader(new FileReader(file));

        List<Integer> list = new ArrayList<>();
        String line;
        while ((line = br.readLine()) != null) {
            list.add(Integer.parseInt(line.trim()));
        }

        int n = list.size();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = list.get(i);
        }

        long startTime = System.nanoTime();
        quickSort(arr, 0, n - 1);
        long endTime = System.nanoTime();

        System.out.println("\nTempo de execução: " + (endTime - startTime) + " ns");

        System.out.println("\nPrimeiros 50 elementos:");
        for (int i = 0; i < 50 && i < arr.length; i++)
            System.out.print(arr[i] + " ");

        System.out.println("\n\nÚltimos 50 elementos:");
        for (int i = Math.max(0, arr.length - 50); i < arr.length; i++)
            System.out.print(arr[i] + " ");
    }

    static void quickSort(int arr[], int lower, int upper) {
        if (lower >= upper)
            return;
        int p = partition(arr, lower, upper);
        quickSort(arr, lower, p - 1);
        quickSort(arr, p + 1, upper);
    }

    private static int partition(int arr[], int lower, int upper) {
        int pivot = arr[upper];
        int j = lower;
        int tmp;
        for (int i = lower; i <= upper; i++) {
            if (arr[i] < pivot) {
                tmp = arr[i];
                arr[i] = arr[j];
                arr[j] = tmp;
                j++;
            }
        }
        tmp = arr[upper];
        arr[upper] = arr[j];
        arr[j] = tmp;

        return j;
    }
}

File: Algoritmos Estudados/QuickSort.py
import time

def partition(arr, low, high):
    i = low - 1
    pivot = arr[high]

    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

if __name__ == '__main__':
    with open('input.txt', 'r') as file:
        arr = [int(line.strip()) for line in file]

    n = len(arr)
    start_time = time.time()
    quick_sort(arr, 0, n - 1)
    end_time = time.time()

    print("Sorted array is:", arr[:50] + ["..."] + arr[-50:])
    print("Time taken:", end_time - start_time, "seconds")
